File: src/main/java/com/github/wgzhao/dbquery/controller/admin/DataSourcesController.java
package com.github.wgzhao.dbquery.controller.admin;

import com.github.wgzhao.dbquery.dto.CommResponse;
import com.github.wgzhao.dbquery.entities.DataSources;
import com.github.wgzhao.dbquery.repo.DataSourceRepo;
import com.github.wgzhao.dbquery.repo.QueryConfigRepo;
import com.github.wgzhao.dbquery.util.DbUtil;
import io.swagger.v3.oas.annotations.OpenAPIDefinition;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequestMapping("${app.api.manage-prefix}/dataSource")
@RequiredArgsConstructor
public class DataSourcesController
{

    private final DataSourceRepo dataSourceRepo;

    private final QueryConfigRepo queryConfigRepo;  // Add this


    @GetMapping
    public List<DataSources> list()
    {
        return dataSourceRepo.findAll();
    }

    @GetMapping("/{id}")
    public DataSources get(@PathVariable("id") String id)
    {
        return dataSourceRepo.findById(id).orElse(null);
    }

    @PostMapping("/testConnection")
    public CommResponse testConnection(@RequestBody DataSources db)
    {
        return DbUtil.testConnect(db);
    }

    @PostMapping
    public DataSources save(@RequestBody DataSources db)
    {
        return dataSourceRepo.save(db);
    }

    @DeleteMapping("/{id}")
    public CommResponse delete(@PathVariable("id") String id)
    {
        // check if it has some query config associated with this data sources
        if (queryConfigRepo.existsByDataSource(id)) {
            return new CommResponse(400, "Cannot delete: Data source is used by one or more query configurations", null);
        }
        if (dataSourceRepo.existsById(id)) {
            dataSourceRepo.deleteById(id);
            return new CommResponse(200, "", null);
        }
        else {
            return new CommResponse(200, "Data source has deleted", null);
        }
    }
}

File: src/main/kotlin/com/github/wgzhao/dbquery/controller/admin/DataSourcesController.kt
package com.github.wgzhao.dbquery.controller.admin

import com.github.wgzhao.dbquery.dto.ApiResponse
import com.github.wgzhao.dbquery.entities.DataSources
import com.github.wgzhao.dbquery.service.DataSourceService
import com.github.wgzhao.dbquery.service.QueryConfigService
import com.github.wgzhao.dbquery.util.DbUtil
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("\${app.api.manage-prefix}/data-sources")
class DataSourcesController(
    private val dataSourceService: DataSourceService,
    private val queryConfigService: QueryConfigService
) {

    @GetMapping
    fun list(): List<DataSources?> {
        return dataSourceService.findAll()
    }

    @GetMapping("/{id}")
    fun get(@PathVariable("id") id: String): DataSources? {
        return dataSourceService.findById(id).orElse(null)
    }

    @PostMapping("/test-connection")
    fun testConnection(@RequestBody db: DataSources): ApiResponse<String?> {
        val res = DbUtil.testConnect(db)
        if (res == "success") {
            return ApiResponse.success<String?>("Connection successful")
        } else {
            return ApiResponse.error<String?>(500, res)
        }
    }

    @PostMapping
    fun save(@RequestBody db: DataSources): DataSources {
        return dataSourceService.save(db)
    }

    @DeleteMapping("/{id}")
    fun delete(@PathVariable("id") id: String): ApiResponse<String?> {
        // check if it has some query config associated with this data sources
        if (queryConfigService.existsByDataSource(id)) {
            return ApiResponse.error<String?>(
                400,
                "Cannot delete: Data source is used by one or more query configurations"
            )
        }
        dataSourceService.deleteById(id)

        return ApiResponse.success<String?>(null)
    }
}
